File: 프로그래머스/3/43164. 여행경로/여행경로.py
from collections import defaultdict
def solution(tickets):
    graph = defaultdict(list)
    
    for a, b in sorted(tickets):
        graph[a].append(b)
    
    answer = []
    
    def dfs(a):
        while graph[a]:
            dfs(graph[a].pop(0))
        answer.append(a)
    dfs("ICN")
    return answer[::-1]

File: 프로그래머스/3/43164. 여행경로/여행경로.js
function solution(tickets) {
    const dfs = (now) => {
        const nextPath = graph.get(now) || [];
        while (nextPath.length !== 0) {
            dfs(nextPath.shift());
        }
        answer.push(now);
    }
    
    
    const graph = new Map();
    
    tickets.sort((a, b) => a[1].localeCompare(b[1]));
    
    for (let i = 0; i < tickets.length; i++) {
        const [a, b] = tickets[i];
        const path = graph.get(a) || [];
        path.push(b);
        graph.set(a, path);
    }
    
    const answer = [];
    dfs("ICN")
    return answer.reverse();
}